src/main.rs

```diff
@@ -0,0 +1,1630 @@
+    TrapHALT = 0x25,  /* halt the program */
+}
+#[derive(Copy, Clone, Debug)]
+enum VMError {
+    InvalidOpCode,
+    InvalidTrapCode,
+    TerminalIOAttributesGet,
+    TerminalIOAttributesSet,
+}
+
+/// The opcodes for the instructions the architecture supports
+#[derive(Debug)]
+enum OpCode {
+    OpBR = 0b0000 << 12,   /* branch */
+    OpADD = 0b0001 << 12,  /* add  */
+    OpLD = 0b0010 << 12,   /* load */
+    OpST = 0b0011 << 12,   /* store */
+    OpJSR = 0b0100 << 12,  /* jump register */
+    OpAND = 0b0101 << 12,  /* bitwise and */
+    OpLDR = 0b0110 << 12,  /* load register */
+    OpSTR = 0b0111 << 12,  /* store register */
+    OpNOT = 0b1001 << 12,  /* bitwise not */
+    OpLDI = 0b1010 << 12,  /* load indirect */
+    OpSTI = 0b1011 << 12,  /* store indirect */
+    OpJMP = 0b1100 << 12,  /* jump */
+    OpLEA = 0b1110 << 12,  /* load effective address */
+    OpTRAP = 0b1111 << 12, /* execute trap */
+}
+
+enum Flag {
+    Positive = 0b001, /* Set when the result of the previous operation was positive */
+    Zero = 0b010,     /* Set when the result of the previous operation was zero */
+    Negative = 0b100, /* Set when the result of the previous operation was negative */
+}
+
+fn extend_sign_for_integer(value: u16, value_bit_count: u16) -> u16 {
+    // If the first bit of the value is negative, because of how two's complement works, we need to extend it with ones unti lwe have 16 bits to preserve the sign
+    // I check if the first bit of the value is one, and if it is I extend it with ones, otherwise with zeroes
+    let is_immediate_negative = (value & (1 << (value_bit_count - 1))) != 0;
+    // If the sign extension is negative, I fill with 11 bits of 1, since 11+5 = 16. Otherwise I don't need to do anything because the filler would be zeroes
+    if is_immediate_negative {
+        // Doing a bitwise or between the sign extension I need and the immediate value I get the immediate with the sign extended
+        // The sign extension will be a series of ones left of an amount of zeroes equivalent to the bits of my current value
+        value | (0xFFFF << value_bit_count)
+    } else {
+        value
+    }
+}
+
+fn disable_input_buffering(original_tio: &mut Termios) -> Result<(), VMError> {
+    tcgetattr(stdin().as_raw_fd(), original_tio).map_err(|_| VMError::TerminalIOAttributesGet)?;
+    let new_tio = original_tio;
+    new_tio.c_lflag &= !ICANON & !ECHO;
+    tcsetattr(stdin().as_raw_fd(), TCSANOW, new_tio)
+        .map_err(|_| VMError::TerminalIOAttributesSet)?;
+    Ok(())
+}
+
+fn restore_input_buffering(original_tio: &mut Termios) -> Result<(), VMError> {
+    tcsetattr(stdin().as_raw_fd(), TCSANOW, original_tio)
+        .map_err(|_| VMError::TerminalIOAttributesSet)
+}
+
+fn handle_interrupt(original_tio: &mut Termios) -> Result<(), VMError> {
+    restore_input_buffering(original_tio).map_err(|_| VMError::TerminalIOAttributesSet)?;
+    exit(-2);
+}
+
+pub struct LC3VM {
```

Comment:
I'm not a big fan of having the entire logic of the vm in the main.rs file.

Ideally the main.rs file should only do parameter parsing/handling and delegate to library functions for domain specific logic.

At the same time, having everything in main.rs makes the review harder. I suggest modularizing it